tty/pc-tty/ttypc_kbd.c

```diff
@@ -316,62 +404,290 @@ static int ttypc_kbd_interrupt(unsigned int n, void *arg)
-			}
+			while (*s && !libtty_putchar(&ttypc->vt->tty, *s++, NULL));
+			continue;
 		}
+
+		mutexUnlock(ttypc->vt->lock);
+		mutexUnlock(ttypc->lock);
 	}
 }
 
 
-int _ttypc_kbd_init(ttypc_t *ttypc)
+/* Waits for keyboard controller status bit with small timeout */
+static int ttypc_kbd_waitstatus(ttypc_t *ttypc, unsigned char bit, unsigned char state)
+{
+	unsigned int i;
+
+	for (i = 0; i < 0xffff; i++)
+		if (!(inb((void *)((uintptr_t)ttypc->kbd + 4)) & ((1 << bit) ^ (state << bit))))
+			return EOK;
+
+	return -ETIMEDOUT;
+}
+
+
+/* Reads a byte from keyboard controller output buffer */
+static int ttypc_kbd_read(ttypc_t *ttypc)
 {
-	ttypc->extended = 0;
-	ttypc->lockst = 0;
-	ttypc->shiftst = 0;
+	int err;
+
+	/* Wait for output buffer not to be empty */
+	if ((err = ttypc_kbd_waitstatus(ttypc, 0, 1)) < 0)
+		return err;
+
+	return inb(ttypc->kbd);
+}
+
+
+/* Writes a byte to keyboard controller input buffer */
+static int ttypc_kbd_write(ttypc_t *ttypc, unsigned char byte)
+{
+	int err;
+
+	/* Wait for input buffer to be empty */
+	if ((err = ttypc_kbd_waitstatus(ttypc, 1, 0)) < 0)
+		return err;
+	outb(ttypc->kbd, byte);
 
-	mutexCreate(&ttypc->rlock);
-	condCreate(&ttypc->rcond);
+	return EOK;
+}
+
+
+/* May not work for PS/2 emulation through USB legacy support */
+int _ttypc_kbd_updateled(ttypc_t *ttypc)
+{
+	int err, ret = 0;
+
+	/* Disable first controller port */
+	outb((void *)((uintptr_t)ttypc->kbd + 4), 0xad);
+
+	do {
+		/* Send update LEDs command */
+		if ((err = ttypc_kbd_write(ttypc, 0xed)) < 0)
+			break;
+
+		/* ACK response */
+		if ((err = ttypc_kbd_read(ttypc)) != 0xfa)
+			break;
+
+		/* Send LEDs state */
+		if ((err = ttypc_kbd_write(ttypc, (ttypc->lockst >> 4) & 0x07)) < 0)
+			break;
+
+		/* ACK response */
+		if ((err = ttypc_kbd_read(ttypc)) != 0xfa)
+			break;
+
+		/* Successfully updated LEDs state */
+		ret = 1;
+	} while (0);
+
+	/* Enable first controller port */
+	outb((void *)((uintptr_t)ttypc->kbd + 4), 0xae);
+
+	return ret;
+}
 
-	/* Attach interrupt and launch interrupt thread */
-	interrupt(ttypc->irq, ttypc_kbd_interrupt, ttypc, ttypc->rcond, &ttypc->inth);
-	beginthread(ttypc_kbd_ctlthr, 1, &ttypc->kbdthr_stack, sizeof(ttypc->kbdthr_stack), (void *)ttypc);
+
+void ttypc_kbd_destroy(ttypc_t *ttypc)
+{
+	resourceDestroy(ttypc->klock);
+	resourceDestroy(ttypc->kcond);
+}
+
+
+int ttypc_kbd_init(ttypc_t *ttypc)
+{
+	int err;
+
+	if ((err = mutexCreate(&ttypc->klock)) < 0)
+		return err;
+
+	if ((err = condCreate(&ttypc->kcond)) < 0) {
+		resourceDestroy(ttypc->klock);
+		return err;
+	}
+
+	/* Attach interrupt */
+	if ((err = interrupt((ttypc->kirq = 1), ttypc_kbd_interrupt, ttypc, ttypc->kcond, &ttypc->kinth)) < 0) {
+		resourceDestroy(ttypc->klock);
+		resourceDestroy(ttypc->kcond);
+		return err;
+	}
+
+	/* Launch keyboard control thread */
+	if ((err = beginthread(ttypc_kbd_ctlthr, 1, &ttypc->kstack, sizeof(ttypc->kstack), (void *)ttypc)) < 0) {
+		resourceDestroy(ttypc->klock);
```

Comment:
You should do `resourceDestroy` on `ttypc->kinth` too

Reply:
Fixed.